bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/dao/impl/PatientDaoImpl.java

```diff
@@ -58,15 +64,27 @@ public PatientDaoImpl(SessionFactory sessionFactory) {
         validateSearchParams(customAttributeFields, programAttributeFieldName, addressFieldName);
 
         ProgramAttributeType programAttributeType = getProgramAttributeType(programAttributeFieldName);
-
-        SQLQuery sqlQuery = new PatientSearchBuilder(sessionFactory)
+        
+        PatientSearchBuilder builder = new PatientSearchBuilder(sessionFactory)
                 .withPatientName(name)
-                .withPatientAddress(addressFieldName, addressFieldValue, addressSearchResultFields)
                 .withPatientIdentifier(identifier, filterOnAllIdentifiers)
                 .withPatientAttributes(customAttribute, getPersonAttributeIds(customAttributeFields), getPersonAttributeIds(patientSearchResultFields))
                 .withProgramAttributes(programAttributeFieldValue, programAttributeType)
-                .withLocation(loginLocationUuid, filterPatientsByLocation)
-                .buildSqlQuery(length, offset);
+                .withLocation(loginLocationUuid, filterPatientsByLocation);
+        
+        if (!i18n.isEnabled()) {
+        	builder.withPatientAddress(addressFieldName, addressFieldValue, addressSearchResultFields);
+        }
+        else { // when i18n is enabled the address is in a list of matched address i18n codes, if any
+        	List<String> codedAddressFieldValues = i18n.getAddressMessageKeysByLikeName(addressFieldValue);
+        	if (CollectionUtils.isEmpty(codedAddressFieldValues)) {
+        		// if no codes could be found then no patients are matched
+        		return Collections.emptyList();
+        	}
+        	builder.withPatientAddressInList(addressFieldName, codedAddressFieldValues, addressSearchResultFields);
+        }
```

Comment:
@angshu I have tried to make this clearer at a higher level of `PatientDaoImpl`. If I18N is disabled then business is as before, otherwise the new `PatientAddressFieldInListQueryHelper` is used through `builder.withPatientAddressInList(...)`. This keeps the overall code closer to what it was.

Also there is a performance enhancement since an empty list is returned right away when no address i18n codes could be found for the input search string. This allows to get rid of the 'unattainable where clause' hack.